src/include/cpu/rpp_cpu_common.hpp

```diff
@@ -151,83 +219,218 @@ inline void rpp_host_rng_xorwow_f32_initialize_multiseed_stream(RpptXorwowState
+    return _mm_sub_ps(*(__m128 *)&pxS, xmm_p1);                                                                                                 // return  *(float *)&out - 1;
 }
 
-inline float rpp_host_rng_xorwow_f32(RpptXorwowState *xorwowState)
+template<typename T>
+inline void rpp_host_rng_xorwow_state_update(T *xorwowState)
 {
     // Save current first and last x-params of xorwow state and compute t
-    uint t  = xorwowState->x[0];
-    uint s  = xorwowState->x[4];
+    Rpp32s t  = xorwowState->x[0];
+    Rpp32s s  = xorwowState->x[4];
     t ^= t >> 2;
     t ^= t << 1;
     t ^= s ^ (s << 4);
 
     // Update all 6 xorwow state params
-    xorwowState->x[0] = xorwowState->x[1];                                              // set new state param x[0]
-    xorwowState->x[1] = xorwowState->x[2];                                              // set new state param x[1]
-    xorwowState->x[2] = xorwowState->x[3];                                              // set new state param x[2]
-    xorwowState->x[3] = xorwowState->x[4];                                              // set new state param x[3]
-    xorwowState->x[4] = t;                                                              // set new state param x[4]
-    xorwowState->counter = (xorwowState->counter + XORWOW_COUNTER_INC) & 0xFFFFFFFF;    // set new state param counter
+    xorwowState->x[0] = xorwowState->x[1];                              // set new state param x[0]
+    xorwowState->x[1] = xorwowState->x[2];                              // set new state param x[1]
+    xorwowState->x[2] = xorwowState->x[3];                              // set new state param x[2]
+    xorwowState->x[3] = xorwowState->x[4];                              // set new state param x[3]
+    xorwowState->x[4] = t;                                              // set new state param x[4]
+    xorwowState->counter = xorwowState->counter + XORWOW_COUNTER_INC;   // set new state param counter
+}
+
+template<typename T>
+inline Rpp32u rpp_host_rng_xorwow_u32(T *xorwowState)
+{
+    // Update xorwow state
+    rpp_host_rng_xorwow_state_update(xorwowState);
+
+    // Return u32 random number
+    return  xorwowState->x[4] + xorwowState->counter;   // return x[4] + counter
+}
+
+template<typename T>
+inline Rpp32f rpp_host_rng_xorwow_f32(T *xorwowState)
+{
+    // Update xorwow state
+    rpp_host_rng_xorwow_state_update(xorwowState);
 
     // Create float representation and return 0 <= outFloat < 1
-    uint out = (XORWOW_EXPONENT_MASK | ((t + xorwowState->counter) & 0x7FFFFF));        // bitmask 23 mantissa bits, OR with exponent
-    float outFloat = *(float *)&out;                                                    // reinterpret out as float
-    return  outFloat - 1;                                                               // return 0 <= outFloat < 1
+    xorwowState->counter &= 0xFFFFFFFF;                                                             // set new state param counter
```

Comment:
why is this required. Looks like it doesn't change the value

Reply:
It was in the reference algorithm too, but yeah its just masking all 32 bits. Removed now.